contracts/wawaka/exchange/contracts/issuer_authority_base.cpp

```diff
@@ -0,0 +1,330 @@
+
+    StringArray serialized_authority_chain;
+    if (! authority_chain.serialize(serialized_authority_chain))
+        return rsp.error("failed to save authority chain; serialization failed");
+
+    if (! issuer_authority_common_store.set(md_authority_chain_key, serialized_authority_chain))
+        return rsp.error("failed to save authority chain; failed to store data");
+
+    CONTRACT_SAFE_LOG(3, "AUTHORITY: %s", (const char*)serialized_authority_chain.c_data());
+
+    // Mark as initialized
+    ww::exchange::exchange_base::mark_initialized();
+
+    // ---------- RETURN ----------
+    return rsp.success(true);
+}
+
+// -----------------------------------------------------------------
+// METHOD: initialize_derived_authority
+//   initialize the key value store for an issuer that derives authority
+//   from another object such as a vetting organization or another issuer
+//
+// JSON PARAMETERS:
+//  asset_authority_chain -- the object that grants issuance
+//    authority to this contract
+//
+// RETURNS:
+//   true
+// -----------------------------------------------------------------
+bool ww::exchange::issuer_authority_base::initialize_derived_authority(
+    const Message& msg,
+    const Environment& env,
+    Response& rsp)
+{
+    ASSERT_SENDER_IS_OWNER(env, rsp);
+    ASSERT_UNINITIALIZED(rsp);
+
+    // Validate the authority given to the contract object
+    ww::exchange::IssuerAuthorityChain authority_chain;
+    if (! msg.get_value("asset_authority_chain", authority_chain))
+        return rsp.error("missing required parameter; asset_authority_chain");
+
+    StringArray verifying_key;
+    if (! ww::exchange::exchange_base::get_verifying_key(verifying_key))
+        return rsp.error("corrupted state; verifying key not found");
+
+    if (! authority_chain.validate(verifying_key))
+        return rsp.error("invalid parameter; authority chain validation failed");
+
+    // Save the serialized authority chain object
+    StringArray serialized_authority_chain;
+    if (! authority_chain.serialize(serialized_authority_chain))
+        return rsp.error("failed to save authority chain; serialization failed");
+
+    if (! issuer_authority_common_store.set(md_authority_chain_key, serialized_authority_chain))
+        return rsp.error("failed to save authority chain; failed to store data");
+
+    // Save the asset type identifier
+    const StringArray asset_type_id(authority_chain.get_string("asset_type_identifier"));
+    if (! issuer_authority_common_store.set(md_asset_type_id_key, asset_type_id))
+        return rsp.error("failed to store the asset type id");
+
+    // Mark as initialized
+    ww::exchange::exchange_base::mark_initialized();
+
+    // ---------- RETURN ----------
+
+    // the authority given to the issuer is only valid if all of the
+    // dependencies have been committed to the ledger
+    if (! authority_chain.add_dependencies_to_response(rsp))
```

Comment:
A diagram with the different contracts, transactions and related dependencies would help understand why such dependencies play a critical role here.

Elaborating further, I'd argue that PDO is quite intuitive, until the dependencies are necessary. We should provide some documentation to guide a developer on when and how to use these.

One last thing. This documentation gap essentially derives from the fact that these dependencies must be coded by a contract developer. So the question is: can we push these down in the stack (i.e., after the contract method returns), and therefore hide them to developers?

Reply:
that document already exists in the exchange documentation for gipsy. that being said, a better description of the state model would be good.